### PR TITLE
Cr 922 locust uac paging

### PR DESCRIPTION
**Pull request by PeterBochel**
This PR is for:  

- Locust worker supports instance number and reads it own part of the event data file.
- Worker sequentially uses UAC's from event data file.
- Updated Readme, to describe use of new environment variables.

Also:

- Fixed deployment manifest so that master and workers can be deployed to GCP
- Fixed expected page titles to match latest.
- Extend list of know pages to match actual 502 errors from Python.

### CODE REVIEW

**Review by philwhiles — APPROVED** [review]

File: locust_tasks/setup.py [hunk 1]
```diff
@@ -2,18 +2,38 @@
 import csv
 import datetime
 import hashlib
+import sys
+import logging
 
 from uuid import uuid4
 from random import randrange
```
Comment (rpcatling) [inline locust_tasks/setup.py]: random now unused - remove

File: locust_tasks/setup.py [hunk 2]
```diff
@@ -2,18 +2,38 @@
...
+def get_next_case():
+    """
+    Gets the next case to be used.
+    :return: Case data, with a UAC that should be in Firestore.
+    """
+
+    global next_case_index
+
+    next_case = cases[next_case_index]
+#    sys.stdout.write('NEXT: %d %s\n' % (next_case_index, next_case['uac']))
```
Comment (rpcatling) [inline locust_tasks/setup.py]: remove ?

File: locust_tasks/setup.py [hunk 3]
```diff
@@ -22,31 +42,108 @@ def setup():
...
-        parameters = pika.URLParameters(RABBITMQ_URL)
-        connection = pika.BlockingConnection(parameters)
-        channel = connection.channel()
+        publish_test_data()
+
+    # Read in section of event data file for the current instance
+    num_event_rows = get_num_event_data_records()
+    (first_record, last_record) = calculate_section_of_event_data_file(num_event_rows)            
+    read_event_data(first_record, last_record)
+    next_case_index = 0
```
Comment (rpcatling) [inline locust_tasks/setup.py]: next_case_index is a local var shadowing the global , so has no effect. remove line.

File: locust_tasks/setup.py [hunk 4]
```diff
@@ -22,31 +42,108 @@ def setup():
...
+
+
+def get_num_event_data_records():
+    """
+    This function reads the event data file to find out how many records it holds.
+    :return: The number of records in the event data file, not including the header line.
+    """
+
+    lines = 0
+    for line in open(FILE_NAME):
```
Comment (rpcatling) [inline locust_tasks/setup.py]: could replace line with _  since variable "line" is unused.

File: locust_tasks/setup.py [hunk 5]
```diff
@@ -22,31 +42,108 @@ def setup():
...
+        sys.exit("ERROR: Event data file too small. There must be at least one worker instance per record in the file") 
+
+    # Calculate range of file owned by this instance
+    records_per_instance = number_records / max_instances
+    first_record = int(records_per_instance * (instance_num-1))
+    last_record = int(records_per_instance * (instance_num)) -1
+
+    logger.info('Instance %d/%d: Event range: %d...%d inclusive from %d records\n' % (instance_num, max_instances, first_record, last_record, number_records))
+
+    return (first_record, last_record)
```
Comment (rpcatling) [inline locust_tasks/setup.py]: pythonic : parenthesis not needed :)

File: locust_tasks/setup.py [hunk 6]
```diff
@@ -22,31 +42,108 @@ def setup():
...
+    Read in the section of the event data file which belongs to the current instance.
+    :param first_record: The first record from the CSV event data file to use. Numbered from 0 after the header line.
+    :param last_record: The final record from the CSV event data file that belongs to the current instance
+    """
+
+    with open(FILE_NAME, 'r') as infile:
+        reader = csv.DictReader(infile)
+        line_number = 0
+        for line in reader:
+            if line_number >= first_record and line_number <= last_record:
```
Comment (rpcatling) [inline locust_tasks/setup.py:113]: pycharm suggests simplification:
`if first_record <= line_number <= last_record:`

File: locust_tasks/locustfile.py [hunk 7]
```diff
@@ -308,6 +318,10 @@ def report_failure(id, resp, task, failure_message, page_content):
 
     resp.failure(f'ID={id} UAC={task.case["uac"]} Status={resp.status_code}: {failure_message}')
     logger.error(f'ID={id} UAC={task.case["uac"]} Status={resp.status_code}: {failure_message}{error_detail}')
+    
+    # Slow down error reporting when things are going wrong (otherwise hundreds of errors are logged in just a few seconds)
+    time.sleep(5.0)
```
Comment (rpcatling) [inline locust_tasks/locustfile.py:324]: please ensure that this 5 second sleep doesn't have adverse affect of recording errors (eg slows down the whole testing)

File: README.md [hunk 8]
```diff
@@ -43,15 +44,24 @@ Before issuing an Kubertetes commands the following substitutions will be needed
 * TARGET\_HOST - This is the IP of the RH start page. It can be found in the browser by looking at the 'census-rh-performance' environment. Then 'Services & Ingress' -> ingress -> Load balancer IP. eg, 'http://http://34.107.206.101'
 * RABBITMQ\_CONNECTION - This is used if you want Locust to populate RH Firestore with test data, otherwise use 'nil'.
 
+To build and publish the docker image CATD recommended (where TAG\_NAME is set to something like "CR-123_V1"):
```
Comment (rpcatling) [inline README.md]: tiny bit more information to clarify the use of TAG_NAME would be helpful.

**Review by rpcatling — CHANGES_REQUESTED** [review]
minor changes requested as discussed.

File: locust_tasks/locustfile.py (hunk 7 above)
Comment (PeterBochel) [reply locust_tasks/locustfile.py:324]: I've hacked the code to force a sleep for one particular UAC. When one of the tasks does the sleep it has no effect on the progress of the other tasks.
Extended comment to note this fact, to hopefully stop people worrying about this in the future.

File: locust_tasks/setup.py (hunk 1 above)
Comment (PeterBochel) [reply locust_tasks/setup.py]: Removed

File: locust_tasks/setup.py (hunk 2 above)
Comment (PeterBochel) [reply locust_tasks/setup.py]: Deleted

File: locust_tasks/setup.py (hunk 3 above)
Comment (PeterBochel) [reply locust_tasks/setup.py]: Removed

File: locust_tasks/setup.py (hunk 4 above)
Comment (PeterBochel) [reply locust_tasks/setup.py]: Done

File: locust_tasks/setup.py (hunk 5 above)
Comment (PeterBochel) [reply locust_tasks/setup.py]: Parenthesis removed

File: locust_tasks/setup.py (hunk 6 above)
Comment (PeterBochel) [reply locust_tasks/setup.py:113]: This change isn't easy to retest so I'll leave as-is unless you feel strongly about it.

File: README.md (hunk 8 above)
Comment (PeterBochel) [reply README.md]: Explanation of tag_name purpose added. 
Please read the new text.

**Review by rpcatling — APPROVED** [review]

**Pull request merged**